### ISSUE TITLE
Large object is allocating by OSHI - Memory issue

### ISSUE THREAD

**Issue by Shikari0744** [oshi/oshi#2476]
Using Dependency i.e., 
 _<dependency>
                <groupId>com.github.oshi</groupId>
                <artifactId>oshi-core</artifactId>
                <version>**5.2.5**</version>
            </dependency>_

**Is this known ?**
```
INFO   | jvm 2    | 2023/10/10 10:00:32 | Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [xxxx.HelloMBean]: Constructor threw exception; nested exception is java.lang.OutOfMemoryError: Cannot allocate 476139 bytes
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:224)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:87)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateBean(AbstractAutowireCapableBeanFactory.java:1326)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	... 61 more
INFO   | jvm 2    | 2023/10/10 10:00:32 | Caused by: java.lang.OutOfMemoryError: Cannot allocate 476139 bytes
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at com.sun.jna.Memory.<init>(Memory.java:116)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at oshi.driver.windows.registry.HkeyPerformanceDataUtil.readPerfDataBuffer(HkeyPerformanceDataUtil.java:278)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at oshi.driver.windows.registry.HkeyPerformanceDataUtil.readPerfDataFromRegistry(HkeyPerformanceDataUtil.java:104)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at oshi.driver.windows.registry.ProcessPerformanceData.buildProcessMapFromRegistry(ProcessPerformanceData.java:69)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at oshi.software.os.windows.WindowsOperatingSystem.queryProcessMapFromRegistry(WindowsOperatingSystem.java:327)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at oshi.util.Memoizer$1.get(Memoizer.java:87)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at oshi.software.os.windows.WindowsOperatingSystem.processMapToList(WindowsOperatingSystem.java:307)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at oshi.software.os.windows.WindowsOperatingSystem.getProcess(WindowsOperatingSystem.java:301)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:211)
INFO   | jvm 2    | 2023/10/10 10:00:32 | 	... 63 more
```

**Comment by Shikari0744** [comment]
Code : -

```
SystemInfo systemInfo = new SystemInfo();
        OperatingSystem operatingSystem = systemInfo.getOperatingSystem();
        int processId = operatingSystem.getProcessId();
```

**Comment by mprins** [comment]
476139 bytes (< .5 MB) doesn't seem like a particularly large amount of memory for a native lib on windows?... 
What happens when you use a current, supported version of oshi?

**Comment by Shikari0744** [comment]
> 476139 bytes (< .5 MB) doesn't seem like a particularly large amount of memory for a native lib on windows?... What happens when you use a current, supported version of oshi?

Yet to try with version _6.4.6_

But as per my understanding, this is giving a false alarm. And my JVM is unexpectedly crashing and restarting again

**Comment by dbwiddis** [comment]
> Is this known ?

As the stack trace indicates, OSHI is attempting to read the process data from the registry at `HKEY_PERFORMANCE_DATA`.  You can read about the structure of the blob being read [here](https://msdn.microsoft.com/en-us/library/windows/desktop/aa373105(v=vs.85).aspx).

Each object being read, representing a process has about 48 bytes of data plus a UTF-16 String (name) that OSHI uses, plus there's some overhead.  So 476139 would mean on the order of 4700 active processes on the system.  This seems unusually large, but not totally unrealistic.

As @mprins indicated, 5.2.5 is over 3 years old.  There have been several improvements to this, most notably https://github.com/oshi/oshi/pull/2075 which was part of 6.2.0, which fixed many issues associated with filling up GC space with discarded memory objects from this particular branch of code, that may be the underlying cause of your OOM.

I see in the trace you are using `getProcess()` to fetch a single process; obviously reading the entire registry blob for all processes may not be the best option.  You can use the `oshi.os.windows.hkeyperfdata` setting to ignore the registry blob and fetch from performance counters instead, probably a better idea if you're only fetching a single process (but I expect a similar total memory allocation if you're reading all processes):
https://github.com/oshi/oshi/blob/a83fe120e7d553f0de1ac85e544adb67fe7e6d1e/oshi-core/src/main/resources/oshi.properties#L62-L73

**Comment by Shikari0744** [comment]
Is OSHI 5.2.5 supports with Windows 11 ? @dbwiddis 
JNA is  : -
<dependency>
            <groupId>net.java.dev.jna</groupId>
            <artifactId>jna</artifactId>
            <version>5.6.0</version>
        </dependency>

**Comment by dbwiddis** [comment]
OSHI 5.2.5 was released 2020-08-30.
Windows 11 was released 2021-10-05.

So by definition OSHI 5.2.5 did not support Windows 11.

I know of one very specific bug upstream in JNA with `Kernel32Util.getLogicalProcessorInformationEx()`, fixed in OSHI 5.6.1 and JNA 5.8.0, so using earlier versions of either dependency is not Windows 11 compatible.  See https://github.com/oshi/oshi/issues/1554

**Comment by Shikari0744** [comment]
Thanks you so much. It helps a lot. We are using oshi from very long time and its been the best open source library I found to monitor resource consumption.

**Comment by dbwiddis** [comment]
> We have been upgrading to OSHI (v6.2.2) and JNA (v5.12.0)

Any reason not to use the newest versions of both?  I'd specifically recommend JNA 5.12.1 over 5.12.0 if there's a reason you're not using 5.13.0.

**Comment by Shikari0744** [comment]
We need to make sure that nothing breaks after upgrading to latest version directly. I needed some necessary fixes which I am getting in JNA 5.12.X and 6.2.X OSHI version. 

I yet to do 1 round of testing with latest JNA and Oshi version.

**Comment by dbwiddis** [comment]
Hey @Shikari0744, I'm going to close this issue as I think I've answered the question and you have a solid path ahead to a newer OSHI version.  Feel free to reopen or create a new issue if you have any more questions.
